### PR TITLE
8246112: Remove build-time and run-time checks for clock_gettime and CLOCK_MONOTONIC

### PR DESCRIPTION
**Pull request by dholmes-ora**
We are now confident that we have build-time and runtime support for clock_gettime and CLOCK_MONOTONIC on all our OpenJDK supported POSIX platforms - see bug report for some more details on different OS. Consequently we can simplify a lot of the code in this area and move common code to os_posix.

As of glibc 2.17 the necessary functions are in glibc rather than librt, but we (Oracle at least) aren't yet in position to set our minimum Linux version to support that. We still have supported platforms at glibc 2.12. So to address that we link librt at build time on Linux. This seems to work find for older and more modern Linuxes and also works for the Apline Linux with Musl variant.

The changes are in layered commits:

Step 1: Remove build time checks. SUPPORTS_CLOCK_MONOTONIC is assumed true and removed.
Step 2: make supports_monotonic_clock always true and so remove checking in OS code
Step 3: Replace gettimeofday by clock_gettime(CLOCK_REALTIME)
Step 4: Move shared time functions to os_posix.cpp
Step 5: Alway link librt on Linux so we don't rely on glibc > 2.17

Testing: tiers 1-3 for functional testing
             built and checked (-Xlog:os) on Linux with glibc 2.12 and 2.17,  macOS 10.13.6 and 10.15.7

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8246112](https://bugs.openjdk.java.net/browse/JDK-8246112): Remove build-time and run-time checks for clock_gettime and CLOCK_MONOTONIC

### Reviewers
 * [Magnus Ihse Bursie](https://openjdk.java.net/census#ihse) (@magicus - **Reviewer**) ⚠️ Review applies to e59676fdec621b3cfe73ef48f0605b16346f5ae2
 * [Erik Joelsson](https://openjdk.java.net/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to e59676fdec621b3cfe73ef48f0605b16346f5ae2
 * [Gerard Ziemski](https://openjdk.java.net/census#gziemski) (@gerard-ziemski - Committer) ⚠️ Review applies to 99ba6528f0c2d8821839df134da3f6dc03f5cdeb
 * [Harold Seigel](https://openjdk.java.net/census#hseigel) (@hseigel - **Reviewer**) ⚠️ Review applies to 99ba6528f0c2d8821839df134da3f6dc03f5cdeb

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2090/head:pull/2090`
`$ git checkout pull/2090`

### CODE REVIEW

File: make/autoconf/flags-ldflags.m4 [hunk 1]
```diff
@@ -108,6 +108,13 @@ AC_DEFUN([FLAGS_SETUP_LDFLAGS_HELPER],
       OS_LDFLAGS_JVM_ONLY="-Wl,-rpath,@loader_path/. -Wl,-rpath,@loader_path/.."
       OS_LDFLAGS="-mmacosx-version-min=$MACOSX_VERSION_MIN"
     fi
+    if test "x$OPENJDK_TARGET_OS" = xlinux; then
+      # Hotspot needs to link librt to get the clock_* functions.
+      # But once our supported minimum build and runtime platform
+      # has glibc 2.17, this can be removed as the functions are
+      # in libc.
+      OS_LDFLAGS="-lrt"
```
Comment (magicus) [inline make/autoconf/flags-ldflags.m4:116]: I believe this should be `OS_LDFLAGS_JVM_ONLY`, right?

**Review by magicus — CHANGES_REQUESTED** [review]
Looks like a good cleanup. But please minimize the use of -lrt to hotspot only as per the comment above.

**Review by erikj79 — APPROVED** [review]
Build changes look good to me.

**Review by magicus — APPROVED** [review]

File: src/hotspot/os/bsd/os_bsd.inline.hpp [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 1999, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1999, 2030, Oracle and/or its affiliates. All rights reserved.
```
Comment (gerard-ziemski) [inline src/hotspot/os/bsd/os_bsd.inline.hpp:2]: Year 2030 ?

File: src/hotspot/os/bsd/os_bsd.cpp [hunk 3]
```diff
@@ -854,45 +825,15 @@ jlong os::javaTimeNanos() {
...
-    jlong result = jlong(tp.tv_sec) * (1000 * 1000 * 1000) + jlong(tp.tv_nsec);
-    return result;
-  } else {
-    timeval time;
-    int status = gettimeofday(&time, NULL);
-    assert(status != -1, "bsd error");
-    jlong usecs = jlong(time.tv_sec) * (1000 * 1000) + jlong(time.tv_usec);
-    return 1000 * usecs;
-  }
+void os::javaTimeNanos_info(jvmtiTimerInfo *info_ptr) {
```
Comment (gerard-ziemski) [inline src/hotspot/os/bsd/os_bsd.cpp:828]: How come we need `os::javaTimeNanos_info` here if there is already one in `os_posix.cpp`?

File: src/hotspot/os/posix/os_posix.hpp [hunk 4]
```diff
@@ -92,22 +92,7 @@ class Posix {
...
 
-#ifdef SUPPORTS_CLOCK_MONOTONIC
-
-private:
-  static bool _supports_monotonic_clock;
-  // These need to be members so we can access them from inline functions
-  static int (*_clock_gettime)(clockid_t, struct timespec *);
-  static int (*_clock_getres)(clockid_t, struct timespec *);
-public:
   static bool supports_monotonic_clock();
```
Comment (gerard-ziemski) [inline src/hotspot/os/posix/os_posix.hpp:95]: Why do we need this API at all now?

thread.cpp uses it here:

      assert(!os::supports_monotonic_clock(),
             "unexpected time moving backwards detected in JavaThread::sleep()");

so we can just remove this usage?

File: src/hotspot/os/posix/os_posix.cpp [hunk 5]
```diff
@@ -1211,28 +1158,13 @@ void os::Posix::init(void) {
 }
 
 void os::Posix::init_2(void) {
-  log_info(os)("Use of CLOCK_MONOTONIC is%s supported",
-               (_clock_gettime != NULL ? "" : " not"));
+  log_info(os)("Use of CLOCK_MONOTONIC is supported");
```
Comment (gerard-ziemski) [inline src/hotspot/os/posix/os_posix.cpp:1161]: We are keeping this output to avoid breaking whomever might be looking for this text?

**Review by gerard-ziemski — CHANGES_REQUESTED** [review]

**Review by gerard-ziemski — APPROVED** [review]

**Comment by tstuefe** [comment]
Builds fine on AIX. I scheduled tests for the coming nights.
About os::javaTimeNanos, not sure if we still need that for AIX, but monotonic clocks gave us a lot of grief in the past on that .. platform and I prefer it to leave that way for now. Investigating if we can use clock_gettime for os::javaTimeNanos like Linux does can be done in a separate RFE.

**Review by hseigel — APPROVED** [review]
Hi David,
The changes look good.  Just a couple of nits.
1. Could you add a comment explaining why AIX and BSD have their own version of javaTimeNanos_info().
2. A few copyrights need to be updated to 2021.
Thanks, Harold

**Comment by tstuefe** [comment]
I don't see any new issues on AIX in our tests with this patch. Thanks for waiting!

**Comment by hseigel** [comment]
Hi David,

I think that os_posix.cpp is a good place to put a comment explaining why AIX and BSD have their own javaTimeNanos_info(). Stating something like "AIX and BSD have their own versions because their implementations of monotonic time have worked on those platforms." seems fine.

Thanks for doing this!
Harold

File: src/hotspot/os/posix/os_posix.cpp [hunk 6]
```diff
@@ -1367,6 +1281,48 @@ void os::Posix::to_RTC_abstime(timespec* abstime, int64_t millis) {
...
 
+// Common (partly) shared time functions
+
+jlong os::javaTimeMillis() {
+  struct timespec ts;
+  int status = clock_gettime(CLOCK_REALTIME, &ts);
+  assert(status == 0, "clock_gettime error: %s", os::strerror(errno));
+  return jlong(ts.tv_sec) * MILLIUNITS +
+    jlong(ts.tv_nsec) / NANOUNITS_PER_MILLIUNIT;
+}
```
Comment (mkraljt) [inline src/hotspot/os/posix/os_posix.cpp:1292]: David,
Great to see simplifications like this happening in OpenJdk code.
- Posix letting you delete platform specific code.
- Always nice to see 'if' branching eliminated to simplify test coverage and runtime.

**Pull request closed without merging**